core/src/core/sumtree.rs

```diff
@@ -0,0 +1,922 @@
+// Copyright 2016 The Grin Developers
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Sum-Merkle Trees
+//!
+//! Generic sum-merkle tree. See `doc/merkle.md` for design and motivation for
+//! this structure. Most trees in Grin are stored and transmitted as either
+//! (a) only a root, or (b) the entire unpruned tree. For these it is sufficient
+//! to have a root-calculating function, for which the `compute_root` function should
+//! be used.
+//!
+//! The output set structure has much stronger requirements, as it is updated
+//! and pruned in place, and needs to be efficiently storable even when it is
+//! very sparse.
+//!
+
+use core::hash::{Hash, Hashed};
+use ser::{self, Readable, Reader, Writeable, Writer};
+use std::collections::HashMap;
+use std::{self, mem, ops};
+
+/// Trait describing an object that has a well-defined sum that the tree can sum over
+pub trait Summable {
+    /// The type of an object's sum
+    type Sum: Clone + ops::Add<Output=Self::Sum> + Readable + Writeable;
+
+    /// Obtain the sum of the object
+    fn sum(&self) -> Self::Sum;
+}
+
+/// An empty sum that takes no space
+#[derive(Copy, Clone)]
+pub struct NullSum;
+impl ops::Add for NullSum {
+    type Output = NullSum;
+    fn add(self, _: NullSum) -> NullSum { NullSum }
+}
+
+impl Readable for NullSum {
+    fn read(_: &mut Reader) -> Result<NullSum, ser::Error> {
+        Ok(NullSum)
+    }
+}
+
+impl Writeable for NullSum {
+    fn write<W: Writer>(&self, _: &mut W) -> Result<(), ser::Error> {
+        Ok(())
+    }
+}
+
+/// Wrapper for a type that allows it to be inserted in a tree without summing
+pub struct NoSum<T>(T);
+impl<T> Summable for NoSum<T> {
+    type Sum = NullSum;
+    fn sum(&self) -> NullSum { NullSum }
+}
+
+#[derive(Clone)]
+enum NodeData<T: Summable> {
+    /// Node with 2^n children which are not stored with the tree
+    Pruned(T::Sum),
+    /// Actual data
+    Leaf(T),
```

Comment:
As discussed on Gitter, I think we want to tree to only deal with hashes to keep it as succinct as possible. A slight improvement would be to be able to take a commitment directly (as it's only 1 byte over a hash) but it may be a premature optimization at this point.